### ISSUE TITLE
Generic build dies in GC on Debian

### ISSUE THREAD

**Issue by geeohgeegeeoh** [RIPE-NCC/rpki-validator-3#145]
I had it die with "Terminating due to java.lang.OutOfMemoryError: GC overhead limit exceeded"

no tuning to the Mx settings, its otherwise unaltered .conf file state I believe. Large memory system host. Bare metal (not virtualized OS)

**Comment by ties** [comment]
What piece of software does the issues occur in? rpki-validator or rpki-rtr-server?

What debian version are you running? And could you also let me know what exact build of the software you are running (preferrably with a link to the binary used)?

**Comment by geeohgeegeeoh** [comment]
Validator: rpki-validator-3.1-2020.01.13.09.31.26

the binary from https://ftp.ripe.net/tools/rpki/validator3/prod/generic/rpki-validator-3.1-2020.01.13.09.31.26-dist.tar.gz

the application.properties has
```
# jvm.mem.initial=1024m       # -Xms jvm option -> initial memory claimed by the jvm
# jvm.mem.maximum=1536m       # -Xmx jvm option -> maximum memory for the jvm
```

I have the as0 TAL, so I have 3500 extra 'things' and a total of around 65000 extra AS0 prefices.

I will try increasing this limit. I feel the nexus between memory and size of the ROA pool (if that is not premature diagnostic) really needs you to be more flexible about memory somehow, because nobody can constrain when a huge number of new ROA/prefixes appear in the Repository structure worldwide

Basically, dying on GC like this is a DOS risk. I can be taken out, by somebody adding state to RPKI

but that is premature diagnostic. It may be something else. It may be a function of my platform, or the AS0 state being mis-coded or something.

**Comment by ties** [comment]
Thanks for the information. Unfortunately I can not reproduce this right now, so let's investigate further.

First of all: How much memory does the machine have? My environment has the ARIN, AS0, localcert tal and runs fine in a Debian VM with 4GB. The number of objects should be comparable:
```
curl http://localhost:8080/api/objects/validated 2>/dev/null | jq '.data.roas | length'
194810
```

We have had issues with debug jdk's using _much_ more memory.  What JDK & debian version are you running?
```
user@stretch:~/rpki-validator-3.1-2020.01.13.09.31.26$ lsb_release -a | grep Description && java -version
No LSB modules are available.
Description:    Debian GNU/Linux 9.9 (stretch)
openjdk version "1.8.0_242"
OpenJDK Runtime Environment (build 1.8.0_242-8u242-b08-1~deb9u1-b08)
OpenJDK 64-Bit Server VM (build 25.242-b08, mixed mode)
```

> Basically, dying on GC like this is a DOS risk. I can be taken out, by somebody adding state to RPKI
> 
> but that is premature diagnostic. It may be something else. It may be a function of my platform, or the AS0 state being mis-coded or something.

There are a number of (unverified) scenario's which could potentially cause DOS across the whole ecosystem. I have noted this one.

**Comment by lolepezy** [comment]
@geeohgeegeeoh 

As a quick solution, just give it more memory. In general, these cases are pretty hard to reproduce, because one can hit some corner case of parallel processes consuming a lot of memory at once. The amount of memory is growing with the number of TAs and, currently, we do not have a way to guarantee that memory consumption is bounded from above for any process within the validator for every configuration of repository and for any object that can be there. So to eliminate the possibility of DOS attacks on the validator we would need to seriously re-write the whole application.

However, we do have changes prepared and tested for reducing memory usage for the validation process itself, they will be released soon.

**Comment by lolepezy** [comment]
Hi,

you can try the version 3.1-2020.05.08.09.26.49 that we just released, it may improve the memory situation.
